Add information on setting up HSM Node Groups for K3s (#43)
Cray-HPE/docs-uan: docs/portal/developer-portal/advanced/Designating_Application_Nodes_for_K3s.md

# Designating Application Nodes for K3s (Technical Preview)

**WARNING**: This feature is currently a Technical Preview. Future releases will streamline these manual configuration steps. Therefore, some of these configuration options may change in future releases.

## Overview

> Before K3s can be enabled to support UAIs on Application nodes, the Application nodes must be labeled in HSM as either `k3s_server` or `k3s_agent` nodes. The UAN Ansible K3s playbooks use these labels to configure them. Nodes labeled `k3s_server` will become K3s control-plane (master) nodes, while nodes labeled `k3s_agant` will become K3s agent (worker) nodes available to host UAIs.

Reviewer: fix typo
```suggestion
Before K3s can be enabled to support UAIs on Application nodes, the Application nodes must be labeled in HSM as either `k3s_server` or `k3s_agent` nodes. The UAN Ansible K3s playbooks use these labels to configure them. Nodes labeled `k3s_server` will become K3s control-plane (master) nodes, while nodes labeled `k3s_agent` will become K3s agent (worker) nodes available to host UAIs.
```

> Before K3s can be enabled to support UAIs on Application nodes, the Application nodes must be labeled in HSM as either `k3s_server` or `k3s_agent` nodes. The UAN Ansible K3s playbooks use these labels to configure them. Nodes labeled `k3s_server` will become K3s control-plane (master) nodes, while nodes labeled `k3s_agent` will become K3s agent (worker) nodes available to host UAIs.

Reviewer 2: ```suggestion
Before K3s can be enabled to support UAIs on Application nodes, the Application nodes must be grouped in HSM as either `k3s_server` or `k3s_agent` nodes. The UAN Ansible K3s playbook use these groups to determine what role they will have. Nodes grouped as `k3s_server` will become K3s control-plane (master) nodes, while nodes grouped as `k3s_agent` will become K3s agent (worker) nodes.
```

> When UAN software is installed or upgraded using IUF and these HSM node group labels don't exist or have no members, one Application_UAN node will be labeled `k3s_server` and the remaining nodes will be labeled `k3s_agent`.  If these groups exist and are not empty, IUF will not change them.

Reviewer: ```suggestion
When UAN software is installed or upgraded using IUF and these HSM node group labels do not exist or have no members, one `Application_UAN` node type will be labeled `k3s_server` and the remaining nodes will be labeled `k3s_agent`.  If these groups exist and are not empty, IUF will not change them.
```

This document provides procedures to manually change the membership of the `k3s_server` and `k3s_agent` HSM node groups.

## Procedures

### Getting a List of all Application_UAN Nodes 

This procedure will gather all nodes with the Application role and UAN subrole from HSM. The output is sorted by XNAME.

```bash
$ cray hsm state components list \
--role Application --subrole UAN \
--format json | jq -r '.Components[].ID' \
| sort
   ```

### Getting the Members of a HSM Node Group

This procedure will display the members of a given HSM Node Group. In the context of this document, `GROUP_NAME` would contain either `k3s_server` or `k3s_agent`.

```bash
$ cray hsm groups members list $GROUP_NAME
```

### Creating the HSM Node Group

> If the HSM Node Group doesn't exist, this procedure may be used to create it. The group members may be provided as a comma-separated string of XNAMEs on the command line, or listed in a file.

Reviewer: ```suggestion
If the HSM Node Group doesn't exist, the `cray hsm groups create` command can create it. The following two examples show how group members may be provided as a comma-separated string of XNAMEs on the command line, or listed in a file.
```

**NOTE:** HSM Node Groups used for K3s must be exclusive.  That is, a given node may be in either the `k3s_server` or `k3s_agent` group, not both. HSM provides the `exclusive-group` element to enforce this policy.  By having the `k3s_server` and `k3s_agent` groups have the same `k3s` `exclusive-group`, nodes are prevented from being members of both.

This example creates the `k3s_server` group listing members on the command line.

 ```bash
 $ cray hsm groups create \
 --members-ids "XNAME1,XNAME2,..." \
 --exclusive-group "K3s" \
 --description "K3s Control-Plane Nodes"\
 --label "k3s_server"
 ```

This example creates the `k3s_agent` group using members in a text file.

 ```bash
 $ cray hsm groups create \
 --members-file /path-to-member-file \
 --exclusive-group "K3s" \
 --description "K3s Agent Nodes"\
 --label "k3s_agent"
   ```

### Modifying the HSM Node Group

> This procedure is used to modify membership of an existing HSM Node Group. In these examples, GROUP_NAME would be either `k3s_server` or `k3s_agent`.

Reviewer: ```suggestion
The `cray hsm groups members create` and `cray hsm groups members delete` commands modify membership of an existing HSM Node Group. In these examples, `GROUP_NAME` would be either `k3s_server` or `k3s_agent`.
```

#### Adding a Member

```bash
$ cray hsm groups members create --id XNAME $GROUP_NAME
```

#### Deleteing a Member

```bash
$ cray hsm groups members delete XNAME $GROUP_NAME
```

### Deleting a HSM Node Group

> This procedure will delete the node group contained in GROUP_NAME.

Reviewer: ```suggestion
This command deletes the node group contained in `GROUP_NAME`.
```
